### DISCUSSION TITLE
2 books + 1 isbn + 2 access numbers = 1 problem CSV import.

### DISCUSSION THREAD

**Discussion by meteos77** [textbrowser/biblioteq#153] [Q&A] [answered]
Context : with age the sight decreases :-) ; a publisher "A vue d'oeil" publishes books with a font size 16.
Consequence: the books are bigger and they are contained in 2 volumes
the 2 volumes have the same ISBN number and almost all the same information
changes in unimarc fields
see the file 978-18466-840-8_differences.txt
[978-2-8466-840-8_différences.txt](https://github.com/textbrowser/biblioteq/files/7780332/978-2-8466-840-8_differences.txt)

[978-2-8466-840-8_volume1.txt](https://github.com/textbrowser/biblioteq/files/7780339/978-2-8466-840-8_volume1.txt)

[978-2-8466-840-8_volume2.txt](https://github.com/textbrowser/biblioteq/files/7780341/978-2-8466-840-8_volume2.txt)

The csv import in BiblioteQ is not possible because the unique key on isbn-13 prevents this

**Comment by textbrowser** [comment]
Yes, that's a problem. Magazines may have identical ISSNs, however, the issue number and volume number are included when defining the primary key.

**Reply by textbrowser** [reply]
Books will need additional keys.

**Reply by textbrowser** [reply]
Or. Add notes stating that this item includes two volumes. Allow the import to fail and modify the entry afterwards.

**Reply by meteos77** [reply]
when you create a book in manual, you can remove the ISBNs check mark, it does not pose a problem?

**Reply by textbrowser** [reply]
Right. There are books which do not have ISBNs.

**Reply by textbrowser** [reply]
So, you can remove the ISBNs from your CSV file and perform the import. Does the import fail is an ISBN is not present?

**Reply by meteos77** [reply]
my work file
[110000-BQ.csv](https://github.com/textbrowser/biblioteq/files/7780920/110000-BQ.csv)

**Reply by textbrowser** [reply]
Great. I can test the import of the enumerations.

**Reply by meteos77** [reply]
in the import I also have a problem with the date of edition ; BiblioteQ wants a DD/MM/YYYY date
all the books I have seen so far only contain the year YYYY

do you print the whole date?

**Reply by meteos77** [reply]
if I delete the isbn of books with several volumes the import goes well

in a database the key "uniq" does not include "not null"?

**Reply by textbrowser** [reply]
The enumerations were created but only 1 item was imported. The IBSNs failed.

**Reply by meteos77** [reply]
new file
[110000-BQ.csv](https://github.com/textbrowser/biblioteq/files/7780948/110000-BQ.csv)

the 1st file is a raw output of the catmandu script, you have to modify the isbn-10 column otherwise it only takes the 1st line

**Reply by textbrowser** [reply]
> if I delete the isbn of books with several volumes the import goes well
> 
> in a database the key "uniq" does not include "not null"?

NULL values are allowed if the NOT NULL is specified. NULLs are unique, even from other NULLs.

**Reply by textbrowser** [reply]
> in the import I also have a problem with the date of edition ; BiblioteQ wants a DD/MM/YYYY date all the books I have seen so far only contain the year YYYY
> 
> do you print the whole date?

I do not have a resolution to this date. The one item which was imported has a publication year of 2012. BQ is not able to parse the year when I open the item's View.

![Screenshot_20211227_113116](https://user-images.githubusercontent.com/10701156/147490375-d4ca1abb-9136-479b-ae90-d1a04670c3ee.png)

**Reply by textbrowser** [reply]
> new file [110000-BQ.csv](https://github.com/textbrowser/biblioteq/files/7780948/110000-BQ.csv)
> 
> the 1st file is a raw output of the catmandu script, you have to modify the isbn-10 column otherwise it only takes the 1st line

I committed the import of enumerations for the book import. Please test it. I will look at the date thing.

**Reply by textbrowser** [reply]
> in the import I also have a problem with the date of edition ; BiblioteQ wants a DD/MM/YYYY date all the books I have seen so far only contain the year YYYY
> 
> do you print the whole date?

The publication date can be any non-null value. However! BQ attempts to display the date in a meaningful format. The preferred format is MM/dd/yyyy. Remember, there is also a display format (Other Options).

**Reply by meteos77** [reply]
Thanks for the reminder (with the numerous options of BiblioteQ, I always miss a good news :-)

in France I do not use the MM/dd/yyyy format at all 
or dd/mm/yyyy
or yyyy-mm-dd for computer filing

I will look in catmandu to output in mm/dd/yyyy but it will be necessary to put 01/01/yyyy because on the unimarc records only the year is marked

**Reply by meteos77** [reply]
after modifying catmandu script it's ok for the date
[unimarc_to_BiblioteQ_import_csv-v0.1.1.fix.zip](https://github.com/textbrowser/biblioteq/files/7781033/unimarc_to_BiblioteQ_import_csv-v0.1.1.fix.zip)

**Reply by textbrowser** [reply]
> > new file [110000-BQ.csv](https://github.com/textbrowser/biblioteq/files/7780948/110000-BQ.csv)
> > the 1st file is a raw output of the catmandu script, you have to modify the isbn-10 column otherwise it only takes the 1st line
> 
> I committed the import of enumerations for the book import. Please test it. I will look at the date thing.

I looked at the new file (110000-BQ.csv), in particular the ISBN-13 values. Most (all) have hyphens and some are ISBN-10 values. Please remember that ISBN-10 is limited to 10 characters and ISBN-13 is limited to 13 characters. Hyphens should not be included nor should ISBN-10 values be present in a field intended for ISBN-13 values.

Then there is this entry. Is this the two-volume entry?

![Screenshot_20211227_215159](https://user-images.githubusercontent.com/10701156/147522727-a001d9e0-2a33-4ac8-a4a5-d28437d5874c.png)

I made the following adjustments for your particular file:
1. Hyphens will be removed from ISBN-10 and ISBN-13 values.
2. If the ISBN-13 is really an ISBN-10, record both the ISBN-10 and the converted ISBN-13.
3. If the ISBN-13 is an ISBN-13 and if the ISBN-10 is not available, create the ISBN-10 from the ISBN-13.

**Reply by textbrowser** [reply]
There is also a problem with all of the ISBN-10 fields. They are all incorrect values. Column 09_ISSN does not include valid ISBN-10 values. My assumption is that you wish BQ to ignore those fields and create the correct values from the ISBN-13 values. This is complicating the import tool.

**Reply by textbrowser** [reply]
Will look more tomorrow if I have time. :)

**Reply by meteos77** [reply]
thanks for looking at my file ; I hope that your gift as a painter is as developed as your programming :-)

indeed, I still have problems with programming the fix script of catmandu with ISBN 10 or 13
the dashes are now removed by the fix command : replace_all(15_isbn13,'[-]','')
fix script modification : [unimarc_to_BiblioteQ_import_csv-v0.1.2.fix.zip](https://github.com/textbrowser/biblioteq/files/7784105/unimarc_to_BiblioteQ_import_csv-v0.1.2.fix.zip)


concerning the line 317, it happens on some records when there are several lines 010 in the unimarc ; see for the record 317
Again, programming problem ; I can't say in script fix to keep only one of the 2 lines

[notice_2-84011-742-8_or_978-2-84011-742-1.txt](https://github.com/textbrowser/biblioteq/files/7783781/notice_2-84011-742-8_or_978-2-84011-742-1.txt)

**Reply by meteos77** [reply]
with version 0.1.4 of the .fix script the isbn-10 and isbn-13 are in their respective fields
when isbn-10 is OK BiblioteQ calculates isbn-13 (SUPER)
if it is the opposite, it refuses the import
[unimarc_to_BiblioteQ_import_csv-v0.1.4.fix.zip](https://github.com/textbrowser/biblioteq/files/7784606/unimarc_to_BiblioteQ_import_csv-v0.1.4.fix.zip)


the file to import
[110000_for_BQ-fix0.1.4.csv.zip](https://github.com/textbrowser/biblioteq/files/7784614/110000_for_BQ-fix0.1.4.csv.zip)

**Reply by textbrowser** [reply]
I just committed my work. I used your latest file (110000_for_BQ-fix0.1.4.csv.zip) as a test. All items, save for five, were imported. The ones which were not imported had NULL ISBN-13s.

**Reply by meteos77** [reply]
GREAT WORK A BIG THANK YOU 
you have improved the importing process, it is going very well

the 5 documents that do not pass are the large print books with 2 volumes instead of one with the same ISBN
in the 1st file there are 5 so 5 imported (the volumes 1) and the 5 errors for the volumes 2

only a human treatment can solve the problem ? 
because when a customer is going to borrow his books they will have 2 access_number to pass to the shower

**Reply by textbrowser** [reply]
> GREAT WORK A BIG THANK YOU you have improved the importing process, it is going very well
> 
> the 5 documents that do not pass are the large print books with 2 volumes instead of one with the same ISBN in the 1st file there are 5 so 5 imported (the volumes 1) and the 5 errors for the volumes 2
> 
> only a human treatment can solve the problem ? because when a customer is going to borrow his books they will have 2 access_number to pass to the shower

The multi-volume sets I've seen have ISBNs assigned to the sets and not to the individual volumes.

**Reply by meteos77** [reply]
ISBN YES hence the conflict

access_number is assigned by the library (2 objects so 2 barcodes)

I put in yellow the duplicates
[BQ_110000-fix0.1.7-color.ods](https://github.com/textbrowser/biblioteq/files/7786023/BQ_110000-fix0.1.7-color.ods)

**Reply by textbrowser** [reply]
I do not mind a new ticket. Simply state the desired changes and we can discuss the potential hazards.

**Comment by meteos77** [comment]
sorting between isbn 10 and isbn13 should be possible with a script fix but then it doesn't make the import in BiblioteQ easier
modification of the right column (10 or 13) ; that's why I leave everything in isbn13

**wait a bit before modifying BiblioteQ**, I'll see what I can do with catmandu

in summary catmandu should do before importing in BiblioteQ
if isbn 10 and isbn13 are present it sorts in the respective fields -> OK with 0.1.3 fix script
if isbn10 is present then it calculates isbn13
if isbn13 is present then it calculates isbn10

**Comment by meteos77** [comment]
the dates are OK 

remaining conversion isbn-13 to isbn-10

the fields "keywords" ; "url" ; are well provided in the import fields in the database but not in the template-1 import csv 
why ?

and the abstract field is not in the list of importable fields

**Reply by textbrowser** [reply]
They're not mandatory fields.

**Reply by meteos77** [reply]
and the abstract field?
because I can import the dimensions of the books
we use it when we look for the books to return (500 / semester)

**Reply by textbrowser** [reply]
If a field is marked as NOT NULL in the SQLite definition, it is required.

**Reply by meteos77** [reply]
by your parser the dimensions of the books are put in the abstract field hence my question

I may have a clue: what is the name of the field in the database? for me it should be abstract and you?

**Reply by textbrowser** [reply]
That's probably out of courtesy. Did you read the release notes?

**Reply by textbrowser** [reply]
I read them and I do not see a mention of the Abstract field and dimensions. So probably out of courtesy. :P

**Reply by meteos77** [reply]
the "description" field in the import tool corresponds to what on the book creation window?

**Reply by meteos77** [reply]
in any case put the release notes in the software (so without the net) it's a great idea !

**Reply by meteos77** [reply]
thanks for the courtesy but we are getting used to having the information :-
![abstract](https://user-images.githubusercontent.com/56431039/147601100-373f1b27-075f-4270-9fe6-88453fe820cb.png)

processing of unimarc line 215
215    $a 1 vol. (276 p.) $c couv. ill. $d 20 cm

**Reply by textbrowser** [reply]
> in any case put the release notes in the software (so without the net) it's a great idea !

Already done. :)

**Reply by meteos77** [reply]
that's why I was congratulating you  it has just appeared, hasn't it?

**Reply by textbrowser** [reply]
> the "description" field in the import tool corresponds to what on the book creation window?

The Abstract. Revised the UI to state Abstract / Description.

**Reply by textbrowser** [reply]
> that's why I was congratulating you it has just appeared, hasn't it?

The release notes were included with the 2021.12.15 release.

![Screenshot_20211228_144458](https://user-images.githubusercontent.com/10701156/147601686-548027e7-caee-4182-bdaa-0002181dc739.png)

**Comment by meteos77** [comment]
I put the current files version 0.1.7
[BQ_110000-fix0.1.7.csv.zip](https://github.com/textbrowser/biblioteq/files/7785943/BQ_110000-fix0.1.7.csv.zip)

catmandu fix script + data (dewey wikipedia)
[fix-0.1.7.zip](https://github.com/textbrowser/biblioteq/files/7785950/fix-0.1.7.zip)

**Comment by textbrowser** [comment]
Submit a ticket indicating which new fields you'd like in the import. Remember, the import is an inverse of the export and the export does not include those fields. :)

**Reply by meteos77** [reply]
before making the ticket on the import
I would need additional fields to fit the 995 and the management of the entries / exits related to our relationship with the Departmental Library

as I know that not many people have our configuration : I submit you my suggestion

create generic fields
for example :  dateperso1 ; dateperso2 (at least 2 date fields) 4 or 6 personal fields : fieldsperso1 ; fieldsperso2 ; fieldsperso3 ; fieldsperso4.

you manage with these names internally and you allow the user to do as with the translation but internally of the software
( for me the field dateperso1 will be presented in the interface as "fields arrival notice") and for someone else ("date of purchase of the book")

**Reply by textbrowser** [reply]
I think we discussed this before; the new fields.

**Comment by meteos77** [comment]
I didn't want to do something to ask you again with the new 995 fields :-)
it's up to you
but I still don't understand how to import in the abstract area (is it possible ?)

**Reply by textbrowser** [reply]
> I didn't want to do something to ask you again with the new 995 fields :-) it's up to you but I still don't understand how to import in the abstract area (is it possible ?)

You can try adding a new row (in the Import tool) and assign it the description field. Then perform the import and see the results. Share the CSV and I'll try.

**Reply by meteos77** [reply]
I have tried or my question on abstract it does not work or I don't know where my mistake is
[BQ_110000-fix0.1.7.csv](https://github.com/textbrowser/biblioteq/files/7786159/BQ_110000-fix0.1.7.csv)
![test-description](https://user-images.githubusercontent.com/56431039/147602015-289e0200-3181-45b9-8c5e-ef9da9aef50f.png)

**Reply by meteos77** [reply]
a track: description is not in the popup either; nor in the table

**Reply by meteos77** [reply]
for me the computer is finished for today good end of day !

**Reply by textbrowser** [reply]
Take care!

**Reply by textbrowser** [reply]
> a track: description is not in the popup either; nor in the table

The description field is automatically set to N/A and your input is ignored. This is logic from the original implementation. So, if the description is not included in the operator's specification, then prepare the default. Otherwise, ignore the default.

**Comment by meteos77** [comment]
import file of the day (some fixes and improvements)  4 new fields
#0.1.8 : added 4 fields : 24_description ; 25_keyword ; 26_marctags ; 27_url

I put the same names as the fields of the BiblioteQ database in the script fix it gives the first line of the csv file with the fields to add at the bottom
[BQ-110000-fix0.1.9.csv.zip](https://github.com/textbrowser/biblioteq/files/7788906/BQ-110000-fix0.1.9.csv.zip)

[Catmandu_fix-0.1.9.zip](https://github.com/textbrowser/biblioteq/files/7788910/Catmandu_fix-0.1.9.zip)

strangely only Abstract/Description is not working?  I made a ticket :-)

**Reply by textbrowser** [reply]
> import file of the day (some fixes and improvements) 4 new fields #0.1.8 : added 4 fields : 24_description ; 25_keyword ; 26_marctags ; 27_url
> 
> I put the same names as the fields of the BiblioteQ database in the script fix it gives the first line of the csv file with the fields to add at the bottom [BQ-110000-fix0.1.9.csv.zip](https://github.com/textbrowser/biblioteq/files/7788906/BQ-110000-fix0.1.9.csv.zip)
> 
> [Catmandu_fix-0.1.9.zip](https://github.com/textbrowser/biblioteq/files/7788910/Catmandu_fix-0.1.9.zip)
> 
> strangely only Abstract/Description is not working? I made a ticket :-)

Committed. Please see if this corrects your import. The description field is now mandatory, although it can be silenced with ```<ignore>```. The three other fields have to be defined in the import process.

**Comment by textbrowser** [comment, marked as answer]
Looking now.

**Reply by meteos77** [reply]
it's ok, thanks I was able to integrate the document summaries in the description field
